### DISCUSSION TITLE
Replacing string with StringSegment or ReadOnlySpan<char>.

### DISCUSSION THREAD

**Discussion by electricessence** [dotnet/runtime#85869] [General]
I've been noticing some places where it seems like `StringSegment` or `ReadOnlySpan<char>` would be much better suited for either parameters or return types.

One example I'd like to ask about is `IConfigurationSection` https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationSection.cs

Seems like it could be beneficial to use something like `StringSegment` for `.Key` or `.Path`.
I can imagine this as a design decision for simplicity, but it seems like a potential memory benefit would be to use something other than potential string portions of the original `string` key path.

Just curious what people's thoughts are if any about this with something like `IConfigurationSection` or any other APIs/Classes where using `StringSegment` could be better.

**Comment by huoyaoyuan** [comment]
It depends on your use pattern. If the string is extracted from a long text, using `StringSegment` will keep the whole original text in memory.

Generally, the short the string lives, it's more beneficial to use span instead of copying.

**Reply by electricessence** [reply]
Yes.  Agreed.  I suggest `StringSegment` as an alternative to `ReadOnlySpan<char>` for the exact reason that in some cases, you will have the original string retained, like the path/key in a dictionary where you only need a 'segment' of that string to identify where you are in the config tree, etc.  I like StringSegment's implicit conversion for parameters since you could reuse a Subsegment as that param.  There's also the obvious downsides of spans as they can't be used with certain scopes or as type params.

**Comment by Joe4evr** [comment]
Changing return/parameter types on shipped APIs is an [unacceptable breaking change](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/breaking-change-rules.md#signatures).

**Reply by electricessence** [reply]
Yes of course.  But we've seen interesting augmentations like `.ValueSpan` exposed in regex captures to allow for better memory use.  You wouldn't change the original API, unless it was 100% compatible, like using `StringSegment` as a param type as it implicitly converts.  What I'm suggesting could easily **_not be_** worth it.  I'm just curious what people think.
